File: algorithms/algorithms-python/leetcode/Question_0067_Add_Binary.py
#!/usr/bin/python
# coding: utf-8

class Solution(object):
    def addBinary(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: str
        """
        if not a:
            return b
        if not b:
            return a
        carry, i, j = 0, len(a) - 1, len(b) - 1
        res = []
        while i >= 0 or j >= 0 or carry == 1:
            if i >= 0:
                carry += int(a[i])
                i -= 1
            if j >= 0:
                carry += int(b[j])
                j -= 1
            res.append(carry % 2)
            carry /= 2
        return "".join([str(i) for i in res[::-1]])

File: algorithms/algorithms-cplusplus/leetcode/Question_0067_Add_Binary.cc
#include <string>

using namespace std;

class Solution {
 public:
  string addBinary(string a, string b) {
    if (a.empty() || b.empty()) {
      return a.empty() ? b : a;
    }
    int carry = 0, i = a.size() - 1, j = b.size() - 1;
    string res = "";
    while (i >= 0 || j >= 0 || carry) {
      carry += i >= 0 ? a[i--] - '0' : 0;
      carry += j >= 0 ? b[j--] - '0' : 0;
      res = char(carry % 2 + '0') + res;
      carry /= 2;
    }
    return res;
  }
};
